modules/service/src/main/scala/lucuma/odb/logic/extensions.scala

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2016-2023 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package lucuma.odb.logic
+
+import cats.Monad
+import cats.data.NonEmptyList
+import cats.syntax.applicative.*
+import cats.syntax.flatMap.*
+import cats.syntax.functor.*
+
+import scala.collection.mutable.ListBuffer
+
+extension [A](as: List[A]) {
+
+  /**
+   * Map over a list, but keeping track of a state when calling the mapping
+   * function whose results are in monadic effect type F.
+   *
+   * @param initialState initial state value
+   * @param f mapping (with state) function
+   * @tparam F effect type
+   * @tparam B resulting element type
+   * @tparam S state type
+   *
+   * @return final state and mapped list
+   */
+  def mapAccumulateM[F[_]: Monad, B, S](initialState: S)(f: (S, A) => F[(S, B)]): F[(S, List[B])] =
```

Comment:
How is this not in cats?

Reply:
I'm probably missing something.  There is

```
def mapAccumulate[S, A, B](init: S, fa: F[A])(f: (S, A) => (S, B)): (S, F[B])
```

in `Traverse` but I didn't see one for a result `F[(S, B)]`.  `fs2.Stream` has `evalMapAccumulate` with that signature that I use elsewhere but I'm not seeing it otherwise.

Reply:
Hmm maybe `as.traverse(a => StateT(s => f(s, a))).run(initialState)`